src/core/android/SDL_android.c

```diff
@@ -2766,4 +2769,85 @@ int Android_JNI_OpenURL(const char *url)
     return ret;
 }
 
+static jstring Android_GetExceptionString(JNIEnv *env)
+{
+    jthrowable throwable = (*env)->ExceptionOccurred(env);
+
+    if (throwable)
+    {
+        (*env)->ExceptionClear(env);
+
+        jclass javaObject = (*env)->FindClass(env, "java/lang/Object");
+        jmethodID toString = (*env)->GetMethodID(env, javaObject, "toString", "()Ljava/lang/String;");
+        jstring message = (*env)->CallObjectMethod(env, throwable, toString);
+
+        (*env)->DeleteLocalRef(env, javaObject);
+        (*env)->DeleteLocalRef(env, throwable);
+        return message;
+    }
+
+    return NULL;
+}
+
+int Android_JNI_OpenFileDescriptor(const char *uri, const char *mode)
+{
+    /* Get fopen-style modes */
+    int moderead = 0, modewrite = 0, modeappend = 0, modeupdate = 0;
+
+    for (const char *cmode = mode; *cmode; cmode++) {
+        switch (*cmode) {
+            case 'a':
+                modeappend = 1;
+                break;
+            case 'r':
+                moderead = 1;
+                break;
+            case 'w':
+                modewrite = 1;
+                break;
+            case '+':
+                modeupdate = 1;
+                break;
+            default:
+                break;
+        }
+    }
+
+    /* Translate fopen-style modes to ContentResolver modes. */
+    /* Android only allows "r", "w", "wt", "wa", "rw" or "rwt". */
+    const char *contentResolverMode = "r";
+
+    if (moderead) {
+        if (modewrite) {
+            contentResolverMode = modeupdate ? "rw" : "rwt";
+        } else {
+            contentResolverMode = modeupdate ? "rw" : "r";
+        }
+    } else if (modewrite) {
+        contentResolverMode = modeupdate ? "rw" : "wt";
+    } else if (modeappend) {
+        contentResolverMode = modeupdate ? "rw" : "wa";
+    }
+
+    JNIEnv *env = Android_JNI_GetEnv();
+    jstring jstringUri = (*env)->NewStringUTF(env, uri);
+    jstring jstringMode = (*env)->NewStringUTF(env, contentResolverMode);
+    jint fd = (*env)->CallStaticIntMethod(env, mActivityClass, midOpenFileDescriptor, jstringUri, jstringMode);
```

Comment:
Maybe do a try/except in Java and return -1 there?
I'm not opposed to this approach, but if we go this route, there are a ton of other Java calls that would benefit from this, and that seems like a separate PR.

Reply:
Sure, I can modify it as needed.

Reply:
Alright I changed it so it handles exception entirely in Java side and return -1 in that case.